### PR TITLE
eth: use the last announced finalized block as the sync ancient limit

### PR DESCRIPTION
**Pull request by karalabe**
Currently we use the hard coded LES CHT as the ancient limit for snap sync. That's unfortunate because a) it introduces a dependency on LES and b) it cannot be updated retrospectively for old clients and c) it needs manual update every release.

This PR adds an additional ancient limit selection logic to the downloader so that if we're in a post-merge network with finalized blocks being announced, then it will use the last finality point as the ancient limit during sync. This mechanism is superior as it will follow the chain dynamically and does not require any manual hardcoded thing.

Note: this PR does not add the same change for normal chain management, as in after initial sync, the ancient limit will revert to 90K and pebble will slowly get filled with ancients. That is a separate concern to fix in a followup PR with it's own implications (reorgs, consensus issues, etc).

### CODE REVIEW

File: eth/downloader/skeleton.go [hunk 1]
```diff
@@ -590,7 +592,17 @@ func (s *skeleton) saveSyncStatus(db ethdb.KeyValueWriter) {
 // accepts and integrates it into the skeleton or requests a reorg. Upon reorg,
 // the syncer will tear itself down and restart with a fresh head. It is simpler
 // to reconstruct the sync state than to mutate it and hope for the best.
-func (s *skeleton) processNewHead(head *types.Header, force bool) bool {
+func (s *skeleton) processNewHead(head *types.Header, final *types.Header, force bool) bool {
+	// If a new finalized block was announced, update the sync process independent
+	// of what happens with the sync head below
+	if final != nil {
+		if number := final.Number.Uint64(); s.progress.Finalized == nil || *s.progress.Finalized != number {
```
Comment (holiman) [inline eth/downloader/skeleton.go:599]: This updates the on disk finalized block with whatever is set, it can go up or down. What happens if we resync the beacon-chain, would it possibly tell is about an older finalized block, and does that matter?

File: eth/downloader/downloader.go [hunk 2]
```diff
@@ -560,26 +560,41 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td, ttd *
...
-			d.ancientLimit = 0
+			// Legacy sync, use any hardcoded checkpoints or the best announcement
+			// we have from the remote peer. TODO(karalabe): Drop this pathway.
+			if d.checkpoint != 0 && d.checkpoint > fullMaxForkAncestry+1 {
+				d.ancientLimit = d.checkpoint
+			} else if height > fullMaxForkAncestry+1 {
+				d.ancientLimit = height - fullMaxForkAncestry - 1
+			} else {
+				d.ancientLimit = 0
+			}
```
Comment (holiman) [inline eth/downloader/downloader.go:597]: how about 

```golang
	if beaconMode && final != nil {
		// Beacon sync, use the latest finalized block as the ancient limit
		// or a reasonable height if no finalized block is yet announced.
		d.ancientLimit = final.Number.Uint64()
	} else if d.checkpoint != 0 && d.checkpoint > fullMaxForkAncestry+1 {
		// Absent finalized blocks, use any hardcoded checkpoints or the best
		// announcement we have from the remote peer.
		d.ancientLimit = d.checkpoint
	} else if height > fullMaxForkAncestry+1 {
		d.ancientLimit = height - fullMaxForkAncestry - 1
	} else {
		d.ancientLimit = 0
	}
```

Comment (karalabe) [reply eth/downloader/downloader.go:597]: Nah, I want the two cases completely separate as they are different sync paths and the meaning of latest is different. Duplicating a bit of code is still better + will be easier to clean once we drop legacy sync altogether.

File: eth/downloader/skeleton.go [hunk 3]
```diff
@@ -437,7 +439,7 @@ func (s *skeleton) sync(head *types.Header) (*types.Header, error) {
 			// we don't seamlessly integrate reorgs to keep things simple. If the
 			// network starts doing many mini reorgs, it might be worthwhile handling
 			// a limited depth without an error.
-			if reorged := s.processNewHead(event.header, event.force); reorged {
+			if reorged := s.processNewHead(event.header, event.final, event.force); reorged {
```
Comment (rjl493456442) [inline eth/downloader/skeleton.go:442]: The final here is always nil, isn't it?

We have two ways to send head event:
- the event sent by `BeaconSync` in order to trigger the initial sync
- the event sent by `BeaconExtend` in order to integrate the newly arrived header

Finalized block will only be provided in `BeaconSync`, the corresponding event is not handled here, but here https://github.com/ethereum/go-ethereum/blob/master/eth/downloader/skeleton.go#L254

In summary, the finalized block will always be nil IIUC.

File: eth/downloader/skeleton.go [hunk 4]
```diff
@@ -1150,38 +1162,45 @@ func (s *skeleton) cleanStales(filled *types.Header) error {
...
 	tail = rawdb.ReadSkeletonHeader(s.db, progress.Subchains[0].Tail)
 	if tail == nil {
-		return nil, nil, fmt.Errorf("tail skeleton header %d is missing", progress.Subchains[0].Tail)
+		return nil, nil, nil, fmt.Errorf("tail skeleton header %d is missing", progress.Subchains[0].Tail)
+	}
+	if progress.Finalized != nil && tail.Number.Uint64() <= *progress.Finalized && *progress.Finalized <= head.Number.Uint64() {
+		final = rawdb.ReadSkeletonHeader(s.db, *progress.Finalized)
+		if final == nil {
+			return nil, nil, nil, fmt.Errorf("finalized skeleton header %d is missing", *progress.Finalized)
+		}
```
Comment (holiman) [inline eth/downloader/skeleton.go:1201]: A new error-mode is introduced here: a missing finalized header will now cause error-return. Previously we didn't require the finalized header to be present -- I don't understand why the skeleton needs the presence of that header.. ?

Also, I don't see why `tail.Number.Uint64() <= *progress.Finalized` is correct. Maybe I just need some :coffee:

Comment (rjl493456442) [reply eth/downloader/skeleton.go:1201]: The error case is: the finalized block is announcement by consensus layer, and also it falls in the range of skeleton. In this case the header should be available in disk. Otherwise error out.

Finalized block is just the *latest finalized block* of the network, but [tail, head] is the range of missing chain segment of local node.

**Review by holiman — APPROVED** [review]
LGTM
